File: Day2/dive.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <string.h>

const char* filename = "input.txt";

void pb1 () {
	FILE* file = fopen(filename, "r");
	assert(file!=NULL);

	char instruc[10];
	int nb, horiz=0, depth=0;
	while (fscanf(file, "%s %i", instruc, &nb) != EOF) {
		if (strcmp(instruc, "forward")==0) {
			horiz += nb;
		} else if (strcmp(instruc, "down")==0) {
			depth += nb;
		} else if (strcmp(instruc, "up")==0) {
			depth -= nb;			
		}
	}


	printf("Résultat pb1 : %i\n", horiz*depth);
	fclose(file);
}

void pb2 () {
	FILE* file = fopen(filename, "r");
	assert(file!=NULL);

	char instruc[10];
	int nb, horiz=0, depth=0, aim=0;
	while (fscanf(file, "%s %i", instruc, &nb) != EOF) {
		if (strcmp(instruc, "forward")==0) {
			horiz += nb;
			depth += nb*aim;
		} else if (strcmp(instruc, "down")==0) {
			aim += nb;
		} else if (strcmp(instruc, "up")==0) {
			aim -= nb;			
		}
	}

	printf("Résultat pb2 : %i\n", horiz*depth);
	fclose(file);
}

int main () {
	pb1();
	pb2();
	return 0;
}

File: Day2/dive.ml
let file_name = "input.txt";;

let dive1 () =
	let file = open_in file_name in
	let rec _dive1 file depth horiz =
		try
			let elts = String.split_on_char ' ' (input_line file) in
			match elts with
			| direc::number::[] -> begin 
				match direc, int_of_string number with
				| "forward", x -> _dive1 file ( depth ) (horiz+x)
				| "down"   , x -> _dive1 file (depth+x) ( horiz )
				| "up"     , x -> _dive1 file (depth-x) ( horiz )
				| _ -> print_endline "FAIL" ; failwith "WTF"
				end
			| _ -> print_endline "FAIL" ; failwith "WTF"
		with e ->
			close_in_noerr file;
			depth * horiz
	in
	_dive1 file 0 0
;;

let dive2 () =
	let file = open_in file_name in
	let rec _dive1 file depth horiz aim =
		try
			let elts = String.split_on_char ' ' (input_line file) in
			match elts with
			| direc::number::[] -> begin 
				match direc, int_of_string number with
				| "forward", x -> _dive1 file (depth+x*aim) (horiz+x) aim
				| "down"   , x -> _dive1 file depth horiz (aim+x)
				| "up"     , x -> _dive1 file depth horiz (aim-x)
				| _ -> print_endline "FAIL" ; failwith "WTF"
				end
			| _ -> print_endline "FAIL" ; failwith "WTF"
		with e ->
			close_in_noerr file;
			depth * horiz
	in
	_dive1 file 0 0 0
;;

let main =
	dive1 () |> Printf.printf "Résultat pb1 : %i\n";
	dive2 () |> Printf.printf "Résultat pb2 : %i\n"
;;
